### ISSUE TITLE
Find Relevant Data

### ISSUE THREAD

**Issue by OmriDan** [OmriDan/NLP_final_project#1]
For testing difficulty estimation on questions, you’ll want datasets that include both the question (and any additional context or answer options) and an associated difficulty label or an indirect proxy (e.g. via performance statistics).

Existing datasets include:

**RACE**: A multiple‐choice reading comprehension dataset split into two subsets (RACE-M for middle school and RACE-H for high school) that naturally provide an approximate difficulty split.
QuAC: While primarily a dialogue-based reading comprehension dataset, protocols have been proposed to label questions as “easy” or “hard” based on multiple QA model performances.
DramaQA: A video QA dataset with manually annotated difficulty levels.
For science, technology, or LeetCode-like programming questions you might consider:

**CodeNet** (IBM): This large-scale dataset of programming problems comes with metadata that can be leveraged or augmented with difficulty labels (if not already provided).
**LeetCode / Competitive Programming Datasets**: Several community repositories on Kaggle aggregate coding problems (often including difficulty ratings as provided on the platform).
**Stack Overflow or Codeforces Data**: Public dumps or curated datasets from these sites can sometimes be re-purposed—by using either user engagement metrics or expert annotation—to serve as proxies for question difficulty.

Note: If a dataset does not natively include difficulty labels, you may derive them via methods such as expert labeling, or by estimating difficulty through Item Response Theory (IRT) using performance data

**Comment by OmriDan** [comment]
LeetCode dataset can be found here:
https://www.kaggle.com/datasets/mohitkumar282/leetcode-questions-dataset

Here is a quick breakdown of the dataset:

Number of Records:
Total questions: 3,306

Column Description:
Question_No: The unique question number on Leetcode.
Question: The title of the question.
Topic_tags: Conceptual topics on which the question is based on (1+,2+ etc implies that there are more topic tags)
**Acceptance**: The acceptance rate of each question (as a percentage).
isPremium: Boolean indicating whether the question requires a premium subscription (True or False).
**Difficulty**: The difficulty level of the question (e.g., Easy, Medium, Hard).
Question_Link: The URL to the question on LeetCode.
Solution: The URL to the solutions provided for the question.

Leetcode dataset with written answers:
https://huggingface.co/datasets/greengerong/leetcode/viewer/default/train?q=3Sum+Closest&views%5B%5D=train&row=15

**Comment by OmriDan** [comment]
Project CodeNet dataset can be found here:
https://github.com/IBM/Project_CodeNet?tab=readme-ov-file#benchmarks

paperswithcode link is here:
https://paperswithcode.com/dataset/project-codenet#:~:text=Project%20CodeNet%20is%20a%20large,one%20of%204000%20coding%20problems.

Project CodeNet is a large-scale dataset with approximately 14 million code samples, each of which is an intended solution to one of 4000 coding problems.

Note it also has difficulty score ('rating') in its metadata:

![Image](https://github.com/user-attachments/assets/5240c1db-64ca-415e-b6da-343fee73ec45)

**Comment by OmriDan** [comment]
APPS (Automated Programming Progress Standard)

The APPS dataset consists of problems collected from different open-access coding websites such as Codeforces, Kattis, and more. The APPS benchmark attempts to mirror how humans programmers are evaluated by posing coding problems in unrestricted natural language and evaluating the correctness of solutions. The problems range in difficulty from introductory to collegiate competition level and measure coding ability as well as problem-solving.

The Automated Programming Progress Standard, abbreviated APPS, consists of 10,000 coding problems in total, with 131,836 test cases for checking solutions and 232,444 ground-truth solutions written by humans. Problems can be complicated, as the average length of a problem is 293.2 words. The data are split evenly into training and test sets, with 5,000 problems each. In the test set, every problem has multiple test cases, and the average number of test cases is 21.2. Each test case is specifically designed for the corresponding problem, enabling us to rigorously evaluate program functionality.

https://github.com/hendrycks/apps?tab=readme-ov-file

**Comment by OmriDan** [comment]
Non STEM QA dataset: DramaQA

https://paperswithcode.com/dataset/dramaqa
The DramaQA focuses on two perspectives: 1) Hierarchical QAs as an evaluation metric based on the cognitive developmental stages of human intelligence. 2) Character-centered video annotations to model local coherence of the story. The dataset is built upon the TV drama "Another Miss Oh" and it contains 17,983 QA pairs from 23,928 various length video clips, with each QA pair belonging to one of four difficulty levels

**Comment by OmriDan** [comment]
https://sites.google.com/site/assistmentsdata/datasets/2012-13-school-data-with-affect
ASSISTments [17] is a dataset of questions mostly about math; differently from the other datasets, no answer options are available. Questions are
not manually calibrated but instead pretested, and the difficulty is continuously
distributed in the range [−5; 5] in a Gaussian-like shape, with µ = −0.32 and
σ = 1.45. The train split contains 6,736 questions, and test 2,245.

**Comment by OmriDan** [comment]
The leetcode data is not equally distributed:

![Image](https://github.com/user-attachments/assets/d979221f-2385-41ce-9541-b3777db4541c)

This resulted in overfit in train and mediocre predictions.

We tried oversampling:

![Image](https://github.com/user-attachments/assets/f5cf800f-c09a-4aaa-8d40-c2d05c982f07)
